datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -645,8 +744,55 @@ pub fn in_list(
     expr: Arc<dyn PhysicalExpr>,
     list: Vec<Arc<dyn PhysicalExpr>>,
     negated: &bool,
+    input_schema: &Schema,
 ) -> Result<Arc<dyn PhysicalExpr>> {
-    Ok(Arc::new(InListExpr::new(expr, list, *negated)))
+    let (cast_expr, cast_list) = in_list_cast(expr, list, input_schema)?;
+    Ok(Arc::new(InListExpr::new(cast_expr, cast_list, *negated)))
+}
+
+fn in_list_cast(
+    expr: Arc<dyn PhysicalExpr>,
+    list: Vec<Arc<dyn PhysicalExpr>>,
+    input_schema: &Schema,
+) -> Result<(Arc<dyn PhysicalExpr>, Vec<Arc<dyn PhysicalExpr>>)> {
+    let expr_type = &expr.data_type(input_schema)?;
+    let list_types: Vec<DataType> = list
+        .iter()
+        .map(|list_expr| list_expr.data_type(input_schema).unwrap())
+        .collect();
+    // TODO in the arrow-rs, should support NULL type to Decimal Data type
+    // TODO support in the arrow-rs, NULL value cast to Decimal Value
+    let result_type = get_coerce_type(expr_type, &list_types);
```

Comment:
The IN operator is a shorthand for multiple OR conditions.
We should get the coerced data type for IN expr.

Reply:
In this issue: https://github.com/apache/arrow-datafusion/issues/2759
The type of list is `BOOLEAN, INT32`, and these two type are not compatible.